proposals/NNNN-hlsl-matrix-type.md

```diff
@@ -0,0 +1,136 @@
+* Matrix multiplication with compatible vectors
+* Resolve overloaded functions taking matrix parameters.
+
+Since matrices are represented as vectors,
+the following limitations of Clang vectors may also be relevant
+
+* Operators don't accept vectors with different element types
+* Boolean vectors are a different type incompatible with other vector types
+* Boolean vectors can only be used as operands to logical operators
+* Matrices and vectors have evaluation kind TEK_Scalar
+
+## Proposed solution
+
+Leverage Clang matrix support by creating HLSL-specific
+matrix types that inherit from the existing matrix extension types.
+These types will have methods enabling the additional features and
+supporting code will get variants to support the additional capability.
+
+### Leveraging Existing Support
+
+The Clang matrix extension provides us with a strong foundation on which to
+build HLSL matrices.
+Existing support gives us the ability to:
+
+* Declare matrices in different scopes (local, global, param, templates, etc)
+* Pass matrices as parameters
+* Return from functions
+* Lower to vector representations
+* Load as column/row major
+* Perform most piecewise operations with splatted scalars (missing divide)
+* Explicit casts between matrices with all compatible element types
+
+Though we want to use the existing support, we don't want to mix the matrix types.
+HLSL should not allow declarations using the Clang `matrix_type` attribute
+and emit an error on the `-fenable-matrix flag`.
+HLSL matrices should be available by default when compiling HLSL without
+any additional parameters.
+
+### HLSL Matrix Type
+
+The Clang HLSL Matrix Type inherits from the Clang matrix extension
+`ConstantMatrixType`.
+For many purposes, it can use the same functionality:
+
+* Reading and writing the type out as element type and dimensions
+* Determining the size of a matrix variable
+* AST traversal and other processing of matrix types
+* Lowering to LLVM vectors
+
+It extends that functionality in the following ways to support HLSL features:
+
+* When HLSL is used, dependent sized matrices will be converted to an HLSL
+  matrix instead of `ConstantMatrixType`.
+* Override dimension checks to limit the sizes to values 1-4.
+* Override the element type check to allow for booleans.
+* Adds point accessors for 
+* Alterations to compatibility checks to allow implicit conversion between
+  matrices of different element types including booleans.
+* Alterations to compatibility checks to allow truncation of larger to smaller
+  dimension matrices.
+
+The HLSL Matrix templated alias is registered as an external source alias.
+This allows the full list of accepted matrix sizes and types to be aliased in
+their short forms in a default header regardless of template or Clang matrix
+support as well as enabling the templated type declarations.
+
+### Mangling
+
+Mangling of Clang matrices is supported for Itanium but not Microsoft-style
+mangling.
```

Comment:
Is your plan here to upgrade/add support for mangling to the base clang matrices support which will be inherited into the HLSL derived type?.. Or are you saying we would only implement MS mangling in the HLSL derived type?